game/minecraft_nbt.ksy

```diff
@@ -0,0 +1,225 @@
+    * something else (especially `image/x-pcx` and `application/octet-stream`),
+      it is most likely already uncompressed.
+
+  The file `output.nbt` generated by one of the above commands can already be
+  processed with this Kaitai Struct specification.
+
+  This spec **only** implements the Java edition format. There is also
+  a [Bedrock edition](https://wiki.vg/NBT#Bedrock_edition) NBT format,
+  which uses little-endian encoding and has a few other differences, but it isn't
+  as popular as the Java edition format.
+
+  **Implementation note:** strings in `TAG_String` are incorrectly decoded with
+  standard UTF-8, while they are encoded in [**Modified UTF-8**](
+    https://docs.oracle.com/javase/8/docs/api/java/io/DataInput.html#modified-utf-8
+  ) (MUTF-8). That's because MUTF-8 is usually not supported natively by most target
+  languages, and thus one must use external libraries to achieve a fully-compliant
+  decoder. But decoding in standard UTF-8 is still better than nothing, and
+  it usually works fine.
+
+  All Unicode codepoints with incompatible representations in MUTF-8 and UTF-8 are
+  U+0000 (_NULL_), U+D800-U+DFFF (_High_ and _Low Surrogates_) and U+10000-U+10FFFF
+  (_Supplementary_ Planes; includes e.g. emoticons, pictograms). When a _MUTF-8_-encoded
+  string is decoded as UTF-8, every such code point will be replaced with
+  2 (for _NULL_ character), 4 (for _Surrogates_) or 6 (for _Supplementary_ characters)
+  "�" symbols (i.e. U+FFFD _Replacement Character_).
+
+  ---
+
+  Sample files:
+
+    * <https://wiki.vg/NBT#Download>
+    * <https://github.com/twoolie/NBT/blob/f9e892e/tests/world_test/data/scoreboard.dat>
+    * <https://github.com/chmod222/cNBT/tree/3f74b69/testdata>
+    * <https://github.com/PistonDevelopers/hematite_nbt/tree/0b85f89/tests>
+doc-ref:
+  - https://wiki.vg/NBT
+  - https://web.archive.org/web/20110723210920/https://www.minecraft.net/docs/NBT.txt
+  - https://minecraft.gamepedia.com/NBT_format
+seq:
+  - size: 0
+    if: root_type == tag::end # always `false`, but it doesn't matter
+    doc: Invoke `valid` check before we start parsing
+  - id: root
+    type: named_tag
+instances:
+  root_type:
+    pos: 0
+    type: u1
+    enum: tag
+    valid: tag::compound
+types:
+  named_tag:
+    -webide-representation: 'TAG_{type}("{name.data}"): {payload:dec}'
+    seq:
+      - id: type
+        type: u1
+        enum: tag
+      - id: name
+        type: tag_string
+        if: not is_tag_end
+      - id: payload
+        type:
+          switch-on: type
+          cases:
+            tag::byte: s1
+            tag::short: s2
+            tag::int: s4
+            tag::long: s8
+            tag::float: f4
+            tag::double: f8
+            tag::byte_array: tag_byte_array
+            tag::string: tag_string
+            tag::list: tag_list
+            tag::compound: tag_compound
+            tag::int_array: tag_int_array
+            tag::long_array: tag_long_array
```

Comment:
BTW, how about introducing a special reserved type that will never be generated, but that can be used for code self-documentation to explicitly mark that this `case` is taken in account (and it may be possible to emit a warning on switches on enums if not all cases are covered)?

Reply:
> In the case of `tag::end` there is also no payload, right? I would add an explicit `if: not is_tag_end` to this field as well, to make that explicit

Good idea. Added in https://github.com/kaitai-io/kaitai_struct_formats/pull/404/commits/3a3d3cca060b9b0ac0d0646a6f8d4f33eef749d2.